staging/src/k8s.io/apiserver/pkg/server/httplog/httplog.go

```diff
@@ -50,12 +50,13 @@ const respLoggerContextKey respLoggerContextKeyType = iota
 // the http.ResponseWriter. We can recover panics from go-restful, and
 // the logging value is questionable.
 type respLogger struct {
-	hijacked       bool
-	statusRecorded bool
-	status         int
-	statusStack    string
-	addedInfo      strings.Builder
-	startTime      time.Time
+	hijacked           bool
+	statusRecorded     bool
+	status             int
+	statusStack        string
+	addedInfo          strings.Builder
+	addedKeyValuePairs []interface{}
```

Comment:
How about limiting the number of memory objects and allocations added by this PR?
We could make the creation of a respLogger initialize addedKeyValuePairs to `make([]interface{}, 0, 20)` (or whatever turns out to be the common case), so that only one memory alloc and one memory object is added (plus the stuff happening in `Log`, where this PR follows the established pattern; the following idea could improve that too, but should probably not be mixed into this PR).  We could even go further an add a `backingKeyValuePairs [20]interface{}` field and initialize addedKeyValuePairs to `backingKeyValuePairs[:0]`, that would add zero memory objects and allocations in the normal case.

Reply:
On second thought, initializing `addedKeyValuePairs` to `backingKeyValuePairs[12:]` and then changing the logic in `Log` that allocates `keysAndValues` to logic that writes to `backingKeyValuePairs[:12]` would be nice and easy.

Reply:
How does `backingKeyValuePairs` work given we have a number of concurrent requests?

Reply:
In general, in the current code we initialize with nil, and the first append grows it. 
Given that currently we only use it in one place, this seems good enough to me - we can optimize it later.

Reply:
Each request gets its own `respLogger`, so each request would have its own `backingKeyValuePairs`.

Reply:
Yes, let's get this interface improvement in ASAP.  The optimizations I suggested are isolated in the implementation.

Reply:
> Each request gets its own respLogger, so each request would have its own backingKeyValuePairs.

In which case I don't understand the value from having it - I would say that you get the value from it if you could share them between calls by not allocating it all the time.

Anyway - yes, let's merge and improve later.

Reply:
@MikeSpreitzer I will open a separate PR with the optimization.